packages/core/redux/waves/thunks.js

```diff
@@ -129,3 +145,110 @@ const sendSignedTransaction = (entry) => async (dispatch, getState) => {
     dispatch(WavesActions.wavesTxSendSignedTransactionError(error))
   }
 }
+
+export const updateWalletBalance = (wallet) => (dispatch) => {
+  getWalletBalances({ wallet })
+    .then((balancesResult) => {
+      try {
+        const newWallet = new WalletModel({
+          ...wallet,
+          balances: {
+            ...wallet.balances,
+            ...formatBalances(BLOCKCHAIN_WAVES, balancesResult),
+          },
+        })
+
+        wavesProvider.subscribe(wallet.address)
+        dispatch({ type: WALLETS_SET, newWallet })
+      } catch (e) {
+        // eslint-disable-next-line no-console
+        console.error(e.message)
+      }
+    })
+    .catch((e) => {
+      // eslint-disable-next-line no-console
+      console.error('call balances from middleware is failed getWalletBalances', e)
+    })
+}
+
+export const enableWaves = () => async (dispatch) => {
+  dispatch(initToken())
+  dispatch(initWalletFromKeys())
+}
+
+const initToken = () => async (dispatch) => {
+
+  const dao = new WavesDAO(WAVES_WAVES_NAME, WAVES_WAVES_SYMBOL, wavesProvider, WAVES_DECIMALS, 'WAVES')
+  dao.watch()
+
+  const waves = await dao.fetchToken()
+  tokenService.registerDAO(waves, dao)
+  dispatch(TokensActions.tokenFetched(waves))
+  dispatch(initWavesAssetTokens(waves))
+}
+
+export const initWavesAssetTokens = (waves: TokenModel) => async (dispatch) => {
+  const assets = await wavesProvider.getAssets()
+  // do not wait until initialized, it is ok to lazy load all the tokens
+  return Promise.all(
+    Object.keys(assets)
+      .map((m) => new WavesDAO(m, m, wavesProvider, assets[m]['decimals'], assets[m]['id'], waves))
+      .map(async (dao) => {
+        try {
+          const token = await dao.fetchToken()
+          tokenService.registerDAO(token, dao)
```

Comment:
Why is it required to pass both `token` and `dao` to `tokenService.registerDAO` if the `token` is got from `dao`?